### PR TITLE
[Feature](Nereids) Support materialized index selection.

### LINKED ISSUE #12303: [Feature] Support rollup and materialized view in nereids planner.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Currently, OLAP table in nereids planner only supports the duplicated data model. Aggregate and Unique data models are not supported.
This issue is for supporting aggregate and unique data models and support rollup and materialized view selection.

Subtasks for the issue:

- [x] Main framework for selecting rollup index with aggregate plan node, pre-aggregation for OLAP scan is turned on by default. #12464 
- [x] Support turning off pre-aggregation for OLAP scan table, e.g., select data without aggregate plan node, select by non-matching aggregate function type, etc. #12785
- [ ] Support all the aggregate column types in the storage engine, such as bitmap, HLL, etc.
- [ ] Support complex expressions in an aggregate function that could use pre-aggregation.
- [x] Support materialized view selection for the duplicated data model. https://github.com/apache/doris/pull/13416
- [ ] Support merge-on-write of the unique key model in nereids planner. (https://cwiki.apache.org/confluence/display/DORIS/DSIP-018%3A+Support+Merge-On-Write+implementation+for+UNIQUE+KEY+data+model)

### Use case

N/A.

### Related issues

No.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by wangshuo128**
# Proposed changes

This PR unified the selection of rollup index and materialized view index into uniform logic, which is called selecting materialized index. 

Main steps:

### Find candidate indexes
1. When aggregate is present, it's handled in `SelectMaterializedIndexWithAggregate`.  The base index and indexes that could use pre-aggregation should be used. The pre-aggregation status is determined by aggregation function, grouping expression, and pushdown predicates.
2. When aggregate is not on top of scan node, it's handled in `SelectMaterializedIndexWithoutAggregate`. The base index and indexes that have all the key columns could be used.

### Filter and order the candidate indexes
1. filter indexes that contain all the required output scan columns.
1. filter indexes that could match prefix index most.
1. order the result index by row count, column count, and index id.

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
4. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
5. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
6. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
7. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java [hunk 1]
```diff
@@ -543,7 +398,7 @@ public CheckContext(LogicalOlapScan scan) {
                             )
                     ));
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java:398]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 2]
```diff
@@ -313,22 +312,13 @@ public PlanFragment visitPhysicalOlapScan(PhysicalOlapScan olapScan, PlanTransla
         tupleDescriptor.setRef(tableRef);
         olapScanNode.setSelectedPartitionIds(olapScan.getSelectedPartitionIds());
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:313]: `olapScanNode` is the plan node in old planner.  `OlapScanNode.setXXX` is to set the result of nereids planner to `OlapScanNode` in old planner.

**Review by 924060929 — DISMISSED** [review]

**Comment by 924060929** [comment]
we should flat filter candidate index to refine code later.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 3]
```diff
@@ -164,30 +164,31 @@ public long getSelectedIndexId() {
...
     public PreAggStatus getPreAggStatus() {
         return preAggStatus;
     }
 
     /**
-     * Should apply {@link SelectRollupWithAggregate} or not.
+     * Should apply {@link SelectMaterializedIndexWithAggregate} or not.
      */
-    public boolean shouldSelectRollup() {
+    public boolean shouldSelectIndex() {
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:178]: This was to do two things:
1. judge `indexSelected` is true or false
2. make sure the keys type of the OLAP table is `DUP_KEYS`, `UNIQUE_KEYS`, or `AGG_KEYS`,  but it should not be `PRIMARY_KEYS`.  Even if it seems `PRIMARY_KEYS` is not used, it's safe to ensure the scope to apply materialized view selection rules is right.

I'd unify `shouldSelectIndex()` and `isIndexSelected()` into one.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithoutAggregate.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            LogicalOlapScan scan,
+            Supplier<Set<Slot>> requiredScanOutputSupplier,
+            Supplier<List<Expression>> predicatesSupplier) {
+        switch (scan.getTable().getKeysType()) {
+            case AGG_KEYS:
+            case UNIQUE_KEYS:
+                OlapTable table = scan.getTable();
+                long baseIndexId = table.getBaseIndexId();
+                int baseIndexKeySize = table.getKeyColumnsByIndexId(table.getBaseIndexId()).size();
+                // No on aggregate on scan.
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithoutAggregate.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/Count.java [hunk 5]
```diff
@@ -30,7 +31,7 @@
 import java.util.stream.Collectors;
 
 /** count agg function. */
-public class Count extends AggregateFunction implements AlwaysNotNullable {
+public class Count extends AggregateFunction implements UnaryExpression, AlwaysNotNullable {
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/Count.java]: fixed

**Review by morrySnow — APPROVED** [review]

**Pull request merged**